File: neb/src/service_status.cc
/*
** Copyright 2009-2013,2015,2019-2020 Centreon
**
** Licensed under the Apache License, Version 2.0 (the "License");
** you may not use this file except in compliance with the License.
** You may obtain a copy of the License at
**
**     http://www.apache.org/licenses/LICENSE-2.0
**
** Unless required by applicable law or agreed to in writing, software
** distributed under the License is distributed on an "AS IS" BASIS,
** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
** See the License for the specific language governing permissions and
** limitations under the License.
**
** For more information : contact@centreon.com
*/

#include "com/centreon/broker/neb/service_status.hh"

#include "com/centreon/broker/database/table_max_size.hh"

using namespace com::centreon::broker;
using namespace com::centreon::broker::neb;

/**************************************
 *                                     *
 *            Public Methods           *
 *                                     *
 **************************************/

/**
 *  @brief Default constructor.
 *
 *  Initialize members to 0, NULL or equivalent.
 */
service_status::service_status(uint32_t type)
    : host_service_status(type),
      last_time_critical(0),
      last_time_ok(0),
      last_time_unknown(0),
      last_time_warning(0),
      service_id(0) {}

/**
 *  @brief Copy constructor.
 *
 *  Copy all members of the given service_status to the current
 *  instance.
 *
 *  @param[in] ss Object to copy.
 */
service_status::service_status(service_status const& ss)
    : host_service_status(ss) {
  _internal_copy(ss);
}

/**
 *  Destructor.
 */
service_status::~service_status() {}

/**
 *  @brief Assignment operator.
 *
 *  Copy all members of the given service_status to the current
 *  instance.
 *
 *  @param[in] ss Object to copy.
 *
 *  @return This object.
 */
service_status& service_status::operator=(service_status const& ss) {
  host_service_status::operator=(ss);
  _internal_copy(ss);
  return *this;
}

/**************************************
 *                                     *
 *           Private Methods           *
 *                                     *
 **************************************/

/**
 *  @brief Copy internal members of the given object to the current
 *         instance.
 *
 *  Make a copy of all internal members defined within service_status to
 *  the current instance. This method is used by the copy constructor
 *  and the assignment operator.
 *
 *  @param[in] ss Object to copy.
 */
void service_status::_internal_copy(service_status const& ss) {
  host_name = ss.host_name;
  last_time_critical = ss.last_time_critical;
  last_time_ok = ss.last_time_ok;
  last_time_unknown = ss.last_time_unknown;
  last_time_warning = ss.last_time_warning;
  service_description = ss.service_description;
  service_id = ss.service_id;
}

/**************************************
 *                                     *
 *           Static Objects            *
 *                                     *
 **************************************/

// Mapping.
mapping::entry const service_status::entries[] = {
    mapping::entry(&service_status::acknowledged, "acknowledged"),
    mapping::entry(&service_status::acknowledgement_type,
                   "acknowledgement_type"),
    mapping::entry(&service_status::active_checks_enabled, "active_checks"),
    mapping::entry(&service_status::check_interval, "check_interval"),
    mapping::entry(&service_status::check_period,
                   "check_period",
                   get_services_col_size(services_check_period)),
    mapping::entry(&service_status::check_type, "check_type"),
    mapping::entry(&service_status::current_check_attempt, "check_attempt"),
    mapping::entry(&service_status::current_state, "state"),
    mapping::entry(&service_status::downtime_depth, "scheduled_downtime_depth"),
    mapping::entry(&service_status::enabled, "enabled"),
    mapping::entry(&service_status::event_handler,
                   "event_handler",
                   get_services_col_size(services_event_handler)),
    mapping::entry(&service_status::event_handler_enabled,
                   "event_handler_enabled"),
    mapping::entry(&service_status::execution_time, "execution_time"),
    mapping::entry(&service_status::flap_detection_enabled, "flap_detection"),
    mapping::entry(&service_status::has_been_checked, "checked"),
    mapping::entry(&service_status::host_id,
                   "host_id",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::host_name, nullptr, 0),
    mapping::entry(&service_status::is_flapping, "flapping"),
    mapping::entry(&service_status::last_check,
                   "last_check",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::last_hard_state, "last_hard_state"),
    mapping::entry(&service_status::last_hard_state_change,
                   "last_hard_state_change",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::last_notification,
                   "last_notification",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::last_state_change,
                   "last_state_change",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::last_time_critical,
                   "last_time_critical",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::last_time_ok,
                   "last_time_ok",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::last_time_unknown,
                   "last_time_unknown",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::last_time_warning,
                   "last_time_warning",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::last_update,
                   "last_update",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::latency, "latency"),
    mapping::entry(&service_status::max_check_attempts, "max_check_attempts"),
    mapping::entry(&service_status::next_check,
                   "next_check",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::next_notification,
                   "next_notification",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::no_more_notifications,
                   "no_more_notifications"),
    mapping::entry(&service_status::notification_number, "notification_number"),
    mapping::entry(&service_status::notifications_enabled, "notify"),
    mapping::entry(&service_status::obsess_over, "obsess_over_service"),
    mapping::entry(&service_status::passive_checks_enabled, "passive_checks"),
    mapping::entry(&service_status::percent_state_change,
                   "percent_state_change"),
    mapping::entry(&service_status::retry_interval, "retry_interval"),
    mapping::entry(&service_status::service_description, nullptr, 0),
    mapping::entry(&service_status::service_id,
                   "service_id",
                   mapping::entry::invalid_on_zero),
    mapping::entry(&service_status::should_be_scheduled, "should_be_scheduled"),
    mapping::entry(&service_status::state_type, "state_type"),
    mapping::entry(&service_status::check_command,
                   "check_command",
                   get_services_col_size(services_check_command)),
    mapping::entry(&service_status::output,
                   "output",
                   get_services_col_size(services_output)),
    mapping::entry(&service_status::perf_data,
                   "perfdata",
                   get_services_col_size(services_perfdata)),
    mapping::entry()};

// Operations.
static io::data* new_service_status() {
  return new service_status;
}
io::event_info::event_operations const service_status::operations = {
    &new_service_status};

File: simu/lua/neb/service_status.lua
local _service_status = {}

local function build(service_id, host_id, metric)
  local now = os.time()
  if not _service_status.start then
    _service_status.start = now
    _service_status.finish = now
  end
  if now > _service_status.finish then
    _service_status.finish = now
  end
  local info = "_" .. service_id .. "_" .. host_id
  local d1 = math.random(0, 100)
  local d2 = math.random(0, 100)
  local d3 = math.random(0, 100)
  local d4 = math.random(0, 100)
  local d5 = math.random(0, 100)
  local retval = {
    category = 1,
    element = 24,
    _type = 65560,
    acknowledged = false,
    acknowledgement_type = 0,
    active_checks = true,
    check_attempt = 1,
    check_command = "OS-Linux-SNMP-Cpu",
    check_interval = 5,
    check_period = "24x7",
    check_type = 0,
    checked = true,
    enabled = true,
    event_handler = "",
    event_handler_enabled = true,
    execution_time = 0.204409,
    flap_detection = true,
    flapping = false,
    host_id = host_id,
    last_check = now,
    last_hard_state = 0,
    last_hard_state_change = 1541149764,
    last_state_change = 1541678384,
    last_time_ok = now,
    last_update = now,
    latency = 0.332,
    max_check_attempts = 3,
    next_check = now + 500,
    no_more_notifications = false,
    notification_number = 0,
    notify = false,
    obsess_over_service = true,
    output = "OK: Message from service status " .. info .."\n",
    passive_checks = false,
    percent_state_change = 0,
    perfdata = "'d1" .. info .. "'=" .. d1 .. "%;;;0;100 "
            .. "'d2" .. info .. "'=" .. d2 .. "%;;;0;100 "
            .. "'d3" .. info .. "'=" .. d3 .. "%;;;0;100 "
            .. "'d4" .. info .. "'=" .. d4 .. "%;;;0;100 "
            .. "'d5" .. info .. "'=" .. d5 .. "%;;;0;100 ",
    retry_interval = 1,
    scheduled_downtime_depth = 0,
    service_id = service_id,
    should_be_scheduled = true,
    state = 0,
    state_type = 1,
  }
  return retval
end

local service_status = {
  name = "Service status",
  -- id: instance id
  -- name: instance name
  -- engine: Monitoring engine name in this instance
  -- pid: Monitoring engine pid
  --
  -- return: a neb::instance event
  build = function (stack, count, conn)
    local metric_count = count.metric
    local service_count = count.service
    local host_count = count.host * count.instance
    broker_log:info(0, "BUILD SERVICE STATUS ; service_count = "
                 .. service_count .. " ; host_count = " .. host_count)
    for metric = 1,metric_count do
      for j = 1,host_count do
        for i = 1,service_count do
          table.insert(
                  stack,
                  build(i + (j - 1) * service_count, j, metric))
        end
      end
    end
    broker_log:info(0, "BUILD SERVICE STATUS => FINISHED")
  end,

  check = function (conn, count)
    local service_count = count.service
    local host_count = count.host * count.instance
    local start = _service_status.start
    if not start then
      return false
    end
    local finish = _service_status.finish
    broker_log:info(0, "CHECK SERVICE STATUS")
    local retval = true
    broker_log:info(0, "SELECT count(*) from data_bin where ctime>=" .. start .. " and ctime <= " .. finish)
    local cursor, error_str = conn["storage"]:execute("SELECT count(*) from data_bin where ctime >=" .. start .. " and ctime <= " .. finish)
    local row = cursor:fetch({}, "a")
    if row then
      if tonumber(row['count(*)']) ~= 5 * service_count * host_count * count.metric then
        broker_log:error(0, "index_data not complete, we have " .. row['count(*)'] .. " rows and not " .. tostring(5 * service_count * host_count * count.metric) .. " as expected")
        retval = false
      end
    end

    if not retval then
      broker_log:info(0, "CHECK SERVICE STATUS => NOT DONE")
    else
      broker_log:info(0, "CHECK SERVICE STATUS => DONE")
    end
    return retval
  end
}

return service_status
